File: Problems/328_Odd_Even_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head==NULL || head->next==NULL||head->next->next==NULL)
        return head;
        ListNode* curr = head->next;
        ListNode* oddEnd = head;
        ListNode* evenEnd = head->next;
        bool isOdd = false;
        while(curr){
            if(isOdd){
                ListNode* nxt = curr->next;
                ListNode* evenStart = oddEnd->next;
                oddEnd->next = curr;
                curr->next = evenStart;
                evenEnd->next = nxt;
                curr = nxt;
                oddEnd = oddEnd->next;
            }else{
                evenEnd = curr;
                curr = curr->next;
            }
            isOdd = !isOdd;
        }       

        return head;
    }

};

File: Problems/328_Odd_Even_Linked_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next or not head.next.next:
            return head
        oddTail = head
        evenHead = head.next
        evenTail = head.next
        
        curr = head.next.next
        isOdd = True
        while curr:
            if isOdd:
                evenTail.next = curr.next
                evenTail = evenTail.next
                oddTail.next = curr
                oddTail = oddTail.next
            curr = curr.next
            isOdd = not isOdd
        
        oddTail.next = evenHead
        
        return head